File: leetcode/python/0028_find_the_index_of_the_first_occurence_in_a_string.py
class Solution(object):
    def strStr(self, haystack, needle):
        """
        :type haystack: str
        :type needle: str
        :rtype: int
        """
        # loop through haystack using i, and needle using j:
        # -> if haystack[i] == needle[j]: j++, start = i
        # -> if haystack[i] != needle[j]: j = 0, i = start
        # -> if j == len(needle): return start
        # return -1
        i, j, start = 0, 0, 0
        while i < len(haystack):
            if haystack[i] == needle[j]:
                if j == 0: start = i
                j += 1
                print(j)
            elif haystack[i] != needle[j]:
                if j > 0: i = start
                j = 0

            if j == len(needle): return start
            i += 1
        
        return -1

File: leetcode/java/0028_find_the_index_of_the_first_occurence_in_a_string.java
class Solution {
    public int strStr(String haystack, String needle) {
        int j = 0, firstOccur = 0;

        for (int i = 0; i < haystack.length(); i++) {
            if (haystack.charAt(i) == needle.charAt(j)) {
                if (j == 0) firstOccur = i;
                j++;
            } else if (haystack.charAt(i) != needle.charAt(j) && j > 0) {
                i = firstOccur;
                j = 0;
            }

            if (j == needle.length()) {
                return firstOccur;
            }
        }

        return -1;
    }
}